lib/RefBackend/RecognizeTorchFallbackPass.cpp

```diff
@@ -0,0 +1,101 @@
+//===- RecognizeTorchFallbackPass.cpp ---------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "npcomp/RefBackend/RefBackend.h"
+#include "PassDetail.h"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyDialect.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyOps.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyDialect.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyOps.h"
+#include "npcomp/Dialect/Refback/IR/RefbackOps.h"
+#include "npcomp/Dialect/Refback/IR/RefbackDialect.h"
+#include "npcomp/Dialect/Torch/IR/OpInterfaces.h"
+#include "npcomp/Dialect/Torch/IR/TorchOps.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "recognize-torch-fallback"
+
+using namespace mlir;
+using namespace mlir::NPCOMP;
+using namespace mlir::NPCOMP::refback;
+using namespace mlir::NPCOMP::Basicpy;
+using namespace mlir::NPCOMP::Torch;
+
+namespace {
+
+class EncapsulateKernelCallOpPattern : public OpRewritePattern<Torch::KernelCallOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(Torch::KernelCallOp kernelCall,
+                                PatternRewriter &rewriter) const override {
+    Operation *kcOp = kernelCall.getOperation();
+    // Change nothing if this KernelCallOp does not descend directly from a FuncOp.
+    if (!isa<FuncOp>(kcOp->getParentOp())) {
+      return failure();
+    }
+    // New Torch fallback region.
+    auto torchfb = rewriter.create<refback::TorchFallbackOp>(
+        kernelCall.getLoc(), kernelCall.results().getType(), kernelCall.args());
+    // Build the region body.
+    rewriter.createBlock(&torchfb.doRegion());
+    SmallVector<Value, 6> encapArgs;
```

Comment:
you still need to verify build_list/etc. here (and fail if the op isn't recognized).

Reply:
That's old code. I pushed a commit that adds a for-loop to check each kernel-call arg for build_list/constant ops (for now).
(Ref: https://github.com/llvm/mlir-npcomp/pull/199/commits/5c51d8cb1d81c713b7743e52c420630c2096d5a0)

Reply:
Ah, I see it now. Could you check it inside this loop? You also need to do something checking tensors/arrays specifically -- one of the operands could be an op that produces a tensor/array.